### PR TITLE
Add precise_time module

### PR DESCRIPTION
**Pull request by margaret**
## Changes
This adds the `precise_time` module which exposes system ticks as an int.

I'm pretty much just going off of `shared-bindings/time/` and https://github.com/adafruit/circuitpython/blob/master/docs/common_hal.md 

I basically copy-pasted the `common-hal/time/__init__.*` to `common-hal/precise_time/__init__.*`. Not sure if there's a way to get the `shared_bindings/precise_time` files to use the ones from `common-hal/time/` instead.

## Testing
See comments. Manually tested only on gemma m0 board (that is the board I have).

## Status 👀 
Addressing review comments.

Edit 2018-06-01: planning to finish this this weekend 🐌 

## Issue
https://github.com/adafruit/circuitpython/issues/519 – More accurate time reporting

### CODE REVIEW

File: shared-bindings/precise_time/__init__.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+//|    :platform: SAMD21
+//|
+//| The `precise_time` module 
+//|
+//| .. method:: monotonic()
+//|
+//|   Returns an always increasing value of time with an unknown reference
+//|   point. Only use it to compare against other values from `monotonic`.
+//|   Unlike time.monotonic, which is a float, the precision of this does not
+//|   degrade over time.
```
Comment (tannewt) [inline shared-bindings/precise_time/__init__.c]: What units is this?

File: shared-bindings/precise_time/__init__.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * THE SOFTWARE.
+ */
+
+#ifndef MICROPY_INCLUDED_SHARED_BINDINGS_PRECISE_TIME__INIT__H
+#define MICROPY_INCLUDED_SHARED_BINDINGS_PRECISE_TIME__INIT__H
+
+#include <stdint.h>
+
+extern uint63_t common_hal_time_monotonic(void)
+extern void common_hal_time_delay_ms(uint32_t)
```
Comment (tannewt) [inline shared-bindings/precise_time/__init__.h]: You'll need semicolons here.

**Comment by margaret** [comment]
```
Adafruit CircuitPython 3.0.0-alpha.6-86-g1b4e064-dirty on 2018-05-15; Adafruit Gemma M0 with samd21e18
>>> 
>>> import precise_time
>>> import time
>>> time.monotonic();precise_time.monotonic()
49.099
49100
>>> time.monotonic();precise_time.monotonic()
52.188
52189
>>> time.monotonic();precise_time.monotonic()
54.439
54440
>>> time.monotonic();precise_time.monotonic()
57.65
57651
>>> time.monotonic();precise_time.monotonic()
67.481
67482
>>> time.monotonic();precise_time.monotonic()
100.681
100682
>>> time.monotonic();precise_time.monotonic()
102.383
102384
>>> time.monotonic();precise_time.monotonic()
105.638
105639
>>> time.monotonic();precise_time.monotonic()
190.729
190730
```

**Comment by margaret** [comment]
Manually tested build at f1ea5d3 `precise_time.sleep` on gemma_m0 board

**Comment by margaret** [comment]
Travis builds are failing with
```
E: Package 'gcc-multilib' has no installation candidate
The command "sudo apt-get install -y python3 gcc-multilib pkg-config libffi-dev libffi-dev:i386 qemu-system" failed and exited with 100 during .
```
Seems like a flake? Can someone 🔄 those?

**Comment by dhalbert** [comment]
The new common-hal routines need to be implemented (or dummied up) for the esp8266 and nrf52 builds as well.

**Comment by dhalbert** [comment]
This is going to be great!

**Comment by tannewt** [comment]
@margaret please mention us when this is ready for another review.

**Comment by margaret** [comment]
@tannewt  @dhalbert  ready for review again. Should I rebase and/or squash?

The two failures were on not nrf builds that passed on HEAD~1. Error is `The command "sudo apt-get install realpath" failed and exited with 100 during .` which seems like a flake.

**Comment by dhalbert** [comment]
I restarted the failed jobs and they ran fine.
I made some inline comments a while ago but I think GitHub didnt show you them because I hadn't started a review. Now I did, see what you think.

File: shared-bindings/precise_time/__init__.c [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+MP_DEFINE_CONST_FUN_OBJ_0(precise_time_monotonic_obj, precise_time_monotonic);
+
+//| .. method:: sleep(milliseconds)
+//|
+//|   Sleep for a given number of milliseconds.
+//|
+//|   :param int seconds: the time to sleep in milliseconds
+//|
+STATIC mp_obj_t precise_time_sleep(mp_obj_t milliseconds_o) {
```
[the remark this answers is not in the document]
Comment (margaret) [reply shared-bindings/precise_time/__init__.c]: That makes sense. I'll rename it to `sleep_ms`.

File: shared-bindings/precise_time/__init__.c [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+//|   Returns an always increasing value of time with an unknown reference
+//|   point. Only use it to compare against other values from `monotonic`.
+//|   Unlike time.monotonic, which is a float, precise_time.monotonic returns an
+//|   int so precision will not degrade over time.
+//|
+//|   :return: the current monotonic time
+//|   :rtype: int
+//|
+STATIC mp_obj_t precise_time_monotonic(void) {
+    return mp_obj_new_int_from_uint(common_hal_precise_time_monotonic());
```
[the remark this answers is not in the document]
Comment (margaret) [reply shared-bindings/precise_time/__init__.c]: We hadn't discussed how to deal with overflow, but these suggestions sound good. I'll work on these items later today.

**Comment by jepler** [comment]
I'd hate to do all this work and end up with something that can't keep precise time indefinitely without needing to use longints.  12.4 days to OverflowError on a non-express board is not a great outcome. 12.4 days until your carefully crafted code becomes markedly slower because it switched to 'long's isn't that hot either.

To that end, I have two general ideas.  Option 1, return a preallocated list of two ints, for example [seconds,microseconds] or [seconds,milliseconds].  Now you're good for 2^30 seconds of runtime, but arithmetic on time values becomes more complicated and needs to allocate lists to do it too.

Option 2, operate in modulo mode, such as modulo 2^30, so that the the monotonic time simply wraps after that many milliseconds.  You need some helper functions for comparing times (e.g., time 0 has to be treated as less than 2^30-1) and adding times (e.g., adding 30 seconds to 2^30-1 seconds has to wrap), but at least it's all on the efficient-to-store integer type.  But you can't say anything about time differences greater than 2^30 milliseconds.

These options are pretty half-baked, but maybe there's a good idea in here waiting to be extracted and implemented.

**Comment by margaret** [comment]
@jepler My initial thought is that something like option 1 would be a nicer interface from the user's POV since they won't have to write their own logic to handle the switch every 12 days

**Comment by tannewt** [comment]
@margaret I think the simplest thing is to not support `precise_time` on non-express builds where long ints aren't supported.

This macro can test for it:
```
#if MICROPY_LONGINT_IMPL == MICROPY_LONGINT_IMPL_NONE
```

**Comment by dhalbert** [comment]
I revisited this again, and found this interesting new accepted PEP, to be implemented in Python 3.7: https://www.python.org/dev/peps/pep-0564/. See below. The new functions defined in this PEP return integer nanosecond values. We could implement some of these: we still have a wraparound or overflow problem, but using these names allows compatibility with regular Python.

We could still have a single `precise_time.time_monotonic()` that could return a tuple of `(seconds, nanoseconds)`, or maybe call it `integer_time.time_monotonic()`.

Quoting from PEP 564: 

> This PEP adds six new functions to the time module:
> 
> - time.clock_gettime_ns(clock_id)
> - time.clock_settime_ns(clock_id, time: int)
> - time.monotonic_ns()
> - time.perf_counter_ns()
> - time.process_time_ns()
> - time.time_ns()
> 
> These functions are similar to the version without the _ns suffix, but return a number of nanoseconds as a Python int.
> 
> For example, time.monotonic_ns() == int(time.monotonic() * 1e9) if monotonic() value is small enough to not lose precision.
> 
> These functions are needed because they may return "large" timestamps, like time.time() which uses the UNIX epoch as reference, and so their float-returning variants are likely to lose precision at the nanosecond resolution.

**Comment by dhalbert** [comment]
Note that 2^30 nanoseconds is about 1 second, so on a non-longint built, the new PEP 564 functions aren't really helpful at all.

**Comment by PaulKierstead** [comment]
It would be really nice to have a function which efficiently returned an int with ms or ns resolution. PEP 564 looks like it would be nice as it is CPython compatible. My only real concern is how efficient is it going to be. Ultimately there is a bunch of cases were basically you want to keep comparing the current time to some known point in time; i.e. have you exceeded the timestamp of your next event time. This is a really common use case. When the events are coming fast and furious, it would be nice if it was highly efficient. I'm inclined to think in addition to the functions which go ahead and allocate at  int object (at considerable expense) it might be nice to have one which just compares your previously allocated int to the current tick count (in some useful unit). I'm just musing there. If that is too painful, then the PEP 564 looks best to me, as it is standard-y. It requires some kind of big int support, but unless you want the handstands that utime when through to handle rollover (and things like utimeq), you need big int support anyway. (or at least 60-odd bit support). So the real decisions, IMO, would be PEP 564 + big int vs some kind of utime system where we get time which rolls-over (and all that entails).

**Comment by tannewt** [comment]
@margaret Thanks again for starting this work. I'm closing this PR because we just merged in `time.monotonic_ns` support which is in CPython 3.7.

**Pull request closed without merging**